### ISSUE TITLE
Error with selecting fit parameters for Payne model

### ISSUE THREAD

**Issue by kennethcarrell** [dnidever/doppler#3]
Hello @dnidever,

When I run doppler with the Payne model option and specify which parameters I want to fit (similar to your second Payne model example in the documentation) I am running into problems.  It works fine if I use the default values of teff,logg,fe_h,alpha_h,rv but if I use other values I get one of a few errors.

If I add one more parameter, say vmacro, or use the exact list in your example (teff,logg,fe_h,alpha_h,c_h,rv), I get the error:
> UnboundLocalError: local variable 'initparlabels' referenced before assignment

that corresponds to line 1434 in the rv.py file.  Finding that line, there are 2 identical "if" conditions in lines 1430 and 1432, and since I give 6 parameters and not 5, neither is True, so the initparlabels variable doesn't get set.

Knowing this, if I fit only 5 parameters, but I do not include rv in the list (teff,logg,fe_h,alpha_h,vmacro), I get the error:
> ValueError: not enough values to unpack (expected 1, got 0)

corresponding to line 1771 in the rv.py file.

If I fit for the parameters teff,logg,c_h,rv,vmacro I get the notification "THIS SHOULD NOT HAPPEN!" and the following:
> \> /Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/payne.py(1517)mklabels()
-> for k,name in enumerate(labelnames):
(Pdb) 

and it waits for a keyboard prompt.

If I remove teff or logg from the list I get an error saying they are required input parameters.

If I replace alpha_h with vmacro everything runs fine, but if I replace it with vmicro it gives me the error:
> ValueError: Each lower bound must be strictly less than each upper bound.

corresponding to line 793 in the least_squares.py file.

If I replace alpha_h with c_h it seems to run fine and produce results.

I suspect one of the two "if" statements should be changed in lines 1430 and 1432, although I'm not positive.

(Also, on a minor, nitpicky note, in line 1407 the comment says you are loading the Cannon models and not the Payne models.)

Thanks.
Kenneth

**Comment by dnidever** [comment]
Thanks for reporting this problem.  Can you show the exact commands that you ran to get the various errors, so I can try to reproduce them?

**Comment by kennethcarrell** [comment]
Here is an example from the command line:

> [~] > doppler spec.fits -v --payne --fitpars teff,logg,fe_h,alpha_h,c_h,rv
spec.fits  S/N= 529.2 
Fitting Payne model to spectrum with parameters: TEFF, LOGG, FE_H, ALPHA_H, C_H, RV
Masked 17 outlier or negative pixels
  TEFF    LOGG   FEH   ALPHAFE    VREL  CCPEAK  CHISQ
   3500   4.80  -1.50   0.00    -20.54   0.13  10.89
   4000   4.80  -1.50   0.00    -26.20   0.21  10.82
   5000   4.60  -1.50   0.00    -28.60   0.43  10.62
   6000   4.40  -1.50   0.00    -28.38   0.33   9.28
   7500   4.00  -1.50   0.00    -35.93   0.16  10.51
  15000   4.00  -1.50   0.00    -45.39   0.15   8.82
  25000   4.00  -1.50   0.00    -41.09   0.16   8.75
  40000   4.00  -1.50   0.00    -48.32   0.18  11.16
   3500   0.50  -1.50   0.00    -29.49   0.13  44.43
   4300   1.00  -1.50   0.00    -29.79   0.33  13.99
   4700   2.00  -1.50   0.00    -29.65   0.35  10.71
   5200   3.00  -1.50   0.00    -28.95   0.35   9.40
   3500   4.80  -0.50   0.00    -24.16   0.12  20.22
   4000   4.80  -0.50   0.00    -28.15   0.31  19.07
   5000   4.60  -0.50   0.00    -29.83   0.50  22.21
   6000   4.40  -0.50   0.00    -29.43   0.48  13.43
   7500   4.00  -0.50   0.00    -28.51   0.18  11.45
  15000   4.00  -0.50   0.00    -45.24   0.15   8.58
  25000   4.00  -0.50   0.00    -40.87   0.16   8.87
  40000   4.00  -0.50   0.00    -47.52   0.17  11.13
   3500   0.50  -0.50   0.00    -29.45   0.16  78.59
   4300   1.00  -0.50   0.00    -31.08   0.35  42.80
   4700   2.00  -0.50   0.00    -30.93   0.42  27.82
   5200   3.00  -0.50   0.00    -30.33   0.47  19.72
   3500   4.80  -1.50   0.30    935.89   0.14  10.74
   4000   4.80  -1.50   0.30    -24.90   0.25  10.80
   5000   4.60  -1.50   0.30    -27.76   0.43  11.24
   6000   4.40  -1.50   0.30    -28.04   0.35   9.38
   7500   4.00  -1.50   0.30    -34.37   0.16  10.52
  15000   4.00  -1.50   0.30    -45.39   0.15   8.81
  25000   4.00  -1.50   0.30    -41.18   0.16   8.76
  40000   4.00  -1.50   0.30    -48.48   0.18  11.15
   3500   0.50  -1.50   0.30    -29.45   0.15  42.72
   4300   1.00  -1.50   0.30    -29.71   0.33  15.37
   4700   2.00  -1.50   0.30    -29.29   0.36  11.64
   5200   3.00  -1.50   0.30    -28.74   0.36  10.00
   3500   4.80  -0.50   0.30    -23.22   0.14  19.94
   4000   4.80  -0.50   0.30    -27.44   0.32  20.30
   5000   4.60  -0.50   0.30    -29.46   0.49  24.95
   6000   4.40  -0.50   0.30    -28.98   0.49  15.33
   7500   4.00  -0.50   0.30    -28.19   0.21  11.99
  15000   4.00  -0.50   0.30    -45.36   0.15   8.57
  25000   4.00  -0.50   0.30    -40.92   0.16   8.88
  40000   4.00  -0.50   0.30    -47.72   0.17  11.13
   3500   0.50  -0.50   0.30    -29.74   0.18  76.74
   4300   1.00  -0.50   0.30    -31.38   0.33  45.61
   4700   2.00  -0.50   0.30    -30.96   0.43  30.22
   5200   3.00  -0.50   0.30    -30.04   0.47  21.56
Initial RV fit:
Teff   =  15000.00 K    
logg   =      4.00      
[Fe/H] =     -0.50      
[alpha/Fe] =      0.30      
Vrel   =    -45.36 +/-  0.503 km/s 
chisq =  8.57
Least Squares RV and stellar parameters:
TEFF   =  16637.22      
LOGG   =      3.86      
FE_H   =      0.12      
ALPHA_H =      0.42      
C_H    =      0.49      
RV     =    -38.55      
chisq =  7.62
Masked 0 discrepant pixels
Doppler failed on spec.fits local variable 'initparlabels' referenced before assignment
Traceback (most recent call last):
  File "/Users/kcarrell/Library/Python/3.8/bin/doppler", line 184, in <module>
    out, model, specm = doppler.fit(spec,payne=payne,fitparams=fitparams,fixparams=fixparams,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/__init__.py", line 21, in fit
    return rv.fit(*args,**kwargs)
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/rv.py", line 3642, in fit
    out = fit_payne(spectrum,model=models,fitparams=fitparams,fixparams=fixparams,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/rv.py", line 1757, in fit_payne
    lsout, lsmodel = fit_lsq_payne(specm,model,initpar=lspars0,fitparams=fitparams,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/rv.py", line 1434, in fit_lsq_payne
    initpar = dict(zip(initparlabels,initpar))
UnboundLocalError: local variable 'initparlabels' referenced before assignment

I get the same error when I run it inside a python script, such as:
`out,model,specm = doppler.rv.fit(spec,verbose=True,payne=True,fitparams=("teff","logg","fe_h","alpha_h","c_h","rv"))`
or
`out,model,specm = doppler.fit(spec,verbose=True,payne=True,fitparams=("teff","logg","fe_h","alpha_h","c_h","rv"))`

It runs fine with the default list of parameters, but specifying others won't work with the Payne model.

**Comment by dnidever** [comment]
Thanks.  I fixd the bug.  Pull in the changes and give it a try.

**Comment by kennethcarrell** [comment]
That solved the not running problem, thanks!  There are still a couple of issues, however.

If I try to fit vmacro or vsini I get 0.0 (it doesn't seem to be attempting a fit at all).

> Initial RV fit:
Teff   =  15000.00 K    
logg   =      4.00      
[Fe/H] =     -0.50      
[alpha/Fe] =      0.30      
Vrel   =    -45.36 +/-  0.503 km/s 
chisq =  8.57
Least Squares RV and stellar parameters:
TEFF   =  16689.60      
LOGG   =      3.86      
FE_H   =      0.07      
ALPHA_H =     -0.01      
C_H    =      0.37      
RV     =    -36.83      
VMACRO =      0.00      
chisq =  7.62
Masked 0 discrepant pixels
Least Squares RV and stellar parameters:
TEFF   =  16529.33      
LOGG   =      3.85      
FE_H   =      0.50      
ALPHA_H =      0.47      
C_H    =      0.50      
RV     =    -31.82      
VMACRO =      0.00      
chisq =  6.12
Fine grid best RV = -28.78 km/s
chisq =  6.12
Final parameters:
TEFF   =  16529.33 +/- 112.983      
LOGG   =      3.85 +/-  0.017      
FE_H   =      0.50 +/-  0.090      
ALPHA_H =      0.47 +/-  0.085      
C_H    =      0.50 +/-  0.128      
RV     =    -28.78 +/-  0.728      
VMACRO =      0.00 +/-  0.000      
Vhelio = -17.74 +/-  0.73 km/s
BC = 11.03 km/s
chisq =  6.12
dt = 18.60 sec.
Writing output to spec_doppler.fits


Also, if I try to include vmicro to the parameter list the code fails with this error:

> [~] > doppler spec.fits -v --payne --fitpars teff,logg,fe_h,alpha_h,c_h,rv,vmicro
spec.fits  S/N= 529.2 
Fitting Payne model to spectrum with parameters: TEFF, LOGG, FE_H, ALPHA_H, C_H, RV, VMICRO
Masked 17 outlier or negative pixels
  TEFF    LOGG   FEH   ALPHAFE    VREL  CCPEAK  CHISQ
   3500   4.80  -1.50   0.00    -20.54   0.13  10.89
   4000   4.80  -1.50   0.00    -26.20   0.21  10.82
   5000   4.60  -1.50   0.00    -28.60   0.43  10.62
   6000   4.40  -1.50   0.00    -28.38   0.33   9.28
   7500   4.00  -1.50   0.00    -35.93   0.16  10.51
  15000   4.00  -1.50   0.00    -45.39   0.15   8.82
  25000   4.00  -1.50   0.00    -41.09   0.16   8.75
  40000   4.00  -1.50   0.00    -48.32   0.18  11.16
   3500   0.50  -1.50   0.00    -29.49   0.13  44.43
   4300   1.00  -1.50   0.00    -29.79   0.33  13.99
   4700   2.00  -1.50   0.00    -29.65   0.35  10.71
   5200   3.00  -1.50   0.00    -28.95   0.35   9.40
   3500   4.80  -0.50   0.00    -24.16   0.12  20.22
   4000   4.80  -0.50   0.00    -28.15   0.31  19.07
   5000   4.60  -0.50   0.00    -29.83   0.50  22.21
   6000   4.40  -0.50   0.00    -29.43   0.48  13.43
   7500   4.00  -0.50   0.00    -28.51   0.18  11.45
  15000   4.00  -0.50   0.00    -45.24   0.15   8.58
  25000   4.00  -0.50   0.00    -40.87   0.16   8.87
  40000   4.00  -0.50   0.00    -47.52   0.17  11.13
   3500   0.50  -0.50   0.00    -29.45   0.16  78.59
   4300   1.00  -0.50   0.00    -31.08   0.35  42.80
   4700   2.00  -0.50   0.00    -30.93   0.42  27.82
   5200   3.00  -0.50   0.00    -30.33   0.47  19.72
   3500   4.80  -1.50   0.30    935.89   0.14  10.74
   4000   4.80  -1.50   0.30    -24.90   0.25  10.80
   5000   4.60  -1.50   0.30    -27.76   0.43  11.24
   6000   4.40  -1.50   0.30    -28.04   0.35   9.38
   7500   4.00  -1.50   0.30    -34.37   0.16  10.52
  15000   4.00  -1.50   0.30    -45.39   0.15   8.81
  25000   4.00  -1.50   0.30    -41.18   0.16   8.76
  40000   4.00  -1.50   0.30    -48.48   0.18  11.15
   3500   0.50  -1.50   0.30    -29.45   0.15  42.72
   4300   1.00  -1.50   0.30    -29.71   0.33  15.37
   4700   2.00  -1.50   0.30    -29.29   0.36  11.64
   5200   3.00  -1.50   0.30    -28.74   0.36  10.00
   3500   4.80  -0.50   0.30    -23.22   0.14  19.94
   4000   4.80  -0.50   0.30    -27.44   0.32  20.30
   5000   4.60  -0.50   0.30    -29.46   0.49  24.95
   6000   4.40  -0.50   0.30    -28.98   0.49  15.33
   7500   4.00  -0.50   0.30    -28.19   0.21  11.99
  15000   4.00  -0.50   0.30    -45.36   0.15   8.57
  25000   4.00  -0.50   0.30    -40.92   0.16   8.88
  40000   4.00  -0.50   0.30    -47.72   0.17  11.13
   3500   0.50  -0.50   0.30    -29.74   0.18  76.74
   4300   1.00  -0.50   0.30    -31.38   0.33  45.61
   4700   2.00  -0.50   0.30    -30.96   0.43  30.22
   5200   3.00  -0.50   0.30    -30.04   0.47  21.56
Initial RV fit:
Teff   =  15000.00 K    
logg   =      4.00      
[Fe/H] =     -0.50      
[alpha/Fe] =      0.30      
Vrel   =    -45.36 +/-  0.503 km/s 
chisq =  8.57
Doppler failed on spec.fits Each lower bound must be strictly less than each upper bound.
Traceback (most recent call last):
  File "/Users/kcarrell/Library/Python/3.8/bin/doppler", line 184, in <module>
    out, model, specm = doppler.fit(spec,payne=payne,fitparams=fitparams,fixparams=fixparams,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/__init__.py", line 21, in fit
    return rv.fit(*args,**kwargs)
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/rv.py", line 3642, in fit
    out = fit_payne(spectrum,model=models,fitparams=fitparams,fixparams=fixparams,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/rv.py", line 1745, in fit_payne
    lsout0, lsmodel0 = fit_lsq_payne(specm,model,initpar=initpar,fitparams=fitparams,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/doppler/rv.py", line 1452, in fit_lsq_payne
    lspars, lscov = curve_fit(sp.model,spec.wave.flatten(),spec.flux.flatten(), #dx_lim=dx_lim,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/dlnpyutils/minpack.py", line 774, in curve_fit
    res = least_squares(func, p0, jac=jac, bounds=bounds, method=method,
  File "/Users/kcarrell/Library/Python/3.8/lib/python/site-packages/dlnpyutils/least_squares.py", line 793, in least_squares
    raise ValueError("Each lower bound must be strictly less than each "
ValueError: Each lower bound must be strictly less than each upper bound.

**Comment by dnidever** [comment]
I made some modifications to help with these issues.  VMICRO is actually not one of the Payne model parameters, that's why it gave the error.  I've added some checking of fitparameter names against the Payne model parameters and it also translates VMICRO->VTURB and VROT->VSINI.

VSINI is now getting non-zero values when I run it, but VTURB and VMACRO are still coming out zeros.  But it's clear that they are being fit.  zero is just the best-fitting value.

It's theoretically very straightforward to add bounaries since that's a standard input to curve_fit.  The hard part is figuring out how to make the inputs easy to use and completely generic.  Maybe it could be similar to the fixparams format.  Something like: --bndpars=TEFF:[5000,6000],LOGG:[2.5,4.5]?
